packages/devtools_app/lib/src/extensions/embedded/_view_web.dart

```diff
@@ -2,19 +2,150 @@
+      ..init();
+  }
+
   @override
   Widget build(BuildContext context) {
-    return const Center(
-      child: Text('TODO implement embedded extension web view'),
+    return Container(
+      color: Theme.of(context).scaffoldBackgroundColor,
+      child: HtmlElementView(
+        viewType: _embeddedExtensionController.viewId,
+      ),
+    );
+  }
+}
+
+class _ExtensionIFrameController extends DisposableController
+    with AutoDisposeControllerMixin {
+  _ExtensionIFrameController(this.embeddedExtensionController);
+
+  final EmbeddedExtensionControllerImpl embeddedExtensionController;
+
+  /// Completes when the extension iFrame has received the first event on the
+  /// 'onLoad' stream.
+  late final Completer<void> _iFrameReady;
+
+  /// Completes when the extension's postMessage handler is ready.
+  ///
+  /// We know this handler is ready when we receive a
+  /// [DevToolsExtensionEventType.pong] event from the
+  /// extension, which it will send in response to a
+  /// [DevToolsExtensionEventType.ping] event sent from DevTools.
+  late final Completer<void> _extensionHandlerReady;
+
+  /// Timer that will poll until [_extensionHandlerReady] is complete or until
+  /// [_pollUntilReadyTimeout] has passed.
+  Timer? _pollForExtensionHandlerReady;
+
+  static const _pollUntilReadyTimeout = Duration(seconds: 10);
+
+  void init() {
+    _iFrameReady = Completer<void>();
+
+    unawaited(
+      embeddedExtensionController.extensionIFrame.onLoad.first.then((_) {
+        _iFrameReady.complete();
+      }),
+    );
+
+    html.window.addEventListener('message', _handleMessage);
+
+    autoDisposeStreamSubscription(
+      embeddedExtensionController.extensionPostEventStream.stream
+          .listen((event) async {
+        await _pingExtensionUntilReady();
+        _postMessage(event);
+      }),
+    );
+  }
+
+  void _postMessage(DevToolsExtensionEvent event) async {
+    await _iFrameReady.future;
+    final message = event.toJson();
+    assert(
+      embeddedExtensionController.extensionIFrame.contentWindow != null,
+      'Something went wrong. The iFrame\'s contentWindow is null after the'
+      ' _iFrameReady future completed.',
+    );
+    embeddedExtensionController.extensionIFrame.contentWindow!.postMessage(
+      message,
+      embeddedExtensionController.extensionUrl,
     );
   }
+
+  void _handleMessage(html.Event e) {
+    if (e is html.MessageEvent) {
+      final extensionEvent = DevToolsExtensionEvent.tryParse(e.data);
+      if (extensionEvent != null) {
+        switch (extensionEvent.type) {
+          case DevToolsExtensionEventType.pong:
+            if (!_extensionHandlerReady.isCompleted) {
+              _extensionHandlerReady.complete();
+            }
+            break;
+          default:
+            notificationService.push(
+              'Unknown event received from extension: ${e.data}',
+            );
+        }
+      }
+    }
+  }
+
+  Future<void> _pingExtensionUntilReady() async {
```

Comment:
Do we want time out and give user friendly message if extension is never ready?

Reply:
added a notification to communicate to the user that something went wrong because the extension is not ready. I also added a TODO to consider giving the user a way to retry or file an issue against the extension package.